src/python/zquantum/core/circuits/_decomposition.py

```diff
@@ -0,0 +1,42 @@
+from typing import Iterable, TypeVar
+
+from typing_extensions import Protocol
+
+OperationType = TypeVar("OperationType")
+
+
+class DecompositionRule(Protocol[OperationType]):
+    def production(self, operation: OperationType) -> Iterable[OperationType]:
+        raise NotImplementedError()
```

Comment:
WDYT about using "the `@abstractmethod` trick" so that codecov doesn't complain?

Reply:
Actually, it still does complain. I already tried this trick (see commit c535a50 and a18321f that reverts it). I think we would need to make the whole protocol an abstract class. Or, we could use `pass` instead of rising `NotImplementedError`. WDYT?

Reply:
The "trick" consists of marking methods with `@abstractmethod` decorator (it raises errors on its own) and removing their implementations, either replacing with `pass` or `...`. Would you try that?

Reply:
Hmm, I thought that the second part is unnecessary, but I'll try doing that!